src/multi.rs

```diff
@@ -31,6 +31,10 @@ use crate::{Error, MultiError};
 pub struct Multi {
     raw: *mut curl_sys::CURLM,
     data: Box<MultiData>,
+
+    // Add extra field so that all MultiWakers can track the dropped state of this Multi.
+    #[cfg(feature = "poll_7_66_0")]
+    dropped: std::sync::Arc<std::sync::Mutex<bool>>,
```

Comment:
I think the best way to structure this would, instead of an `Arc<Mutex<T>>` to do something like:

* Create a `RawMulti` wrapper which isn't `pub` which contains the actual `CURLM` pointer. 
* This `RawMulti` wrapper would have `unsafe impl Send/Sync`.
* The `RawMulti` wrapper would have a destructor
* The `Multi` struct would store `Arc<RawMulti>` plus the `data` field it has today.
* The `MultiWaker` struct would store `Weak<RawMulti>`

I think that would handle all the synchronization fiddly bits here and handle destruction pretty gracefully as well.